pennylane/devices/qubit/sampling.py

```diff
@@ -53,47 +123,60 @@ def measure_with_samples(
     Returns:
         TensorLike[Any]: Sample measurement results
     """
-    # if the measurement process involves a Sum or a Hamiltonian, measure each
-    # of the terms separately and sum
-    if isinstance(mp, ExpectationMP):
-        if isinstance(mp.obs, Hamiltonian):
-
-            def _sum_for_single_shot(s):
-                return sum(
-                    c
-                    * measure_with_samples(
-                        ExpectationMP(t), state, s, is_state_batched=is_state_batched, rng=rng
+    groups, indices = _group_measurements(mps)
+
+    all_res = []
+    for group in groups:
+        if isinstance(group[0], ExpectationMP) and isinstance(group[0].obs, Hamiltonian):
```

Comment:
Another optional comment that could be done in a follow-up PR.

This section's starting to get a bit overly nested.

So I'm trying to figure out if there's anything we could do to cut down on some of the nesting.

One potential idea: We could make `_measure_hamiltonian_with_samples`, `_measure_sum_with_samples`, `_measure_classical_shadow` and `_measure_with_samples_diagonalizing_gates` all obey the same call signature.  Then have the `elif-elif` block just choose which method. Then below the `elif-elif` block, we can call the relevant function and stick its results into `all_res`.

Also, if we used `extend` here instead of `append`, we might be able to avoid the flattening step below.